### DISCUSSION TITLE
enabling core library desugaring in expo android app

### DISCUSSION THREAD

**Discussion by saman3d** [expo/expo#31071] [Archived: Questions]
we are using a library which requires core library desugaring to be enabled.
```Dependency 'org.signal:libsignal-android:0.55.1' requires core library desugaring to be enabled```
as i read through the [docs](https://developer.android.com/studio/write/java8-support.html#library-desugaring) `coreLibraryDesugaringEnabled` flag should be set to enabled in compileOptions and 'com.android.tools:desugar_jdk_libs:1.1.9' should be added to dependencies.
now that both are added to gradle configuration of my own library i get a second error regarding my own library requires core library desugaring to be enabled:
```
       1.  Dependency ':react-native-libsignal-client' requires core library desugaring to be enabled
           for :app.
     
           See https://developer.android.com/studio/write/java8-support.html for more
           details.
     
       2.  Dependency 'org.signal:libsignal-android:0.55.1' requires core library desugaring to be enabled
           for :app.
     
           See https://developer.android.com/studio/write/java8-support.html for more
           details.
```

**Comment by long-vv** [comment]
Hi, I'm getting error `requires core library desugaring to be enabled for :app` with another package too
Please help, how can I enable that core library with [expo-build-properties](https://docs.expo.dev/versions/latest/sdk/build-properties/)?

**Comment by anhtuan219** [comment]
expo-build-properties can't be used for this case since it is used to change tool versions like kotlin version, SDK version,..., not to make specific custom native changes
To make custom changes to the native files (e.g. `build.gradle`), you must use [config plugin](https://docs.expo.dev/config-plugins/introduction/)

At the root folder of project, create folder `plugins`, inside that folder, create file `withCustomAppBuildGradle.js` with the content below

```
const { withAppBuildGradle } = require('expo/config-plugins');

module.exports = function withCustomAppBuildGradle(config) {
 return withAppBuildGradle(config, async (config) => {
   const androidPattern = '\nandroid {\n';
   const androidIndex = config.modResults.contents.indexOf(androidPattern);
   const androidPivot = androidIndex + androidPattern.length + 1;
   config.modResults.contents =
     config.modResults.contents.slice(0, androidPivot) +
     '   compileOptions {\n        coreLibraryDesugaringEnabled true\n    }\n\n ' +
     config.modResults.contents.slice(androidPivot);

   const dependenciesPattern = '\ndependencies {\n';
   const dependenciesIndex = config.modResults.contents.indexOf(dependenciesPattern);
   const dependenciesPivot = dependenciesIndex + dependenciesPattern.length + 1;
   config.modResults.contents =
     config.modResults.contents.slice(0, dependenciesPivot) +
     '   coreLibraryDesugaring("com.android.tools:desugar_jdk_libs:2.0.4")\n\n ' +
     config.modResults.contents.slice(dependenciesPivot);

   return config;
 });
};
```

Modify file `app.json` to use config plugin `withCustomAppBuildGradle`

```
{
    "expo": {
        "plugins": [
            "./plugins/withCustomAppBuildGradle.js"
        ]
    }
}
```

Run the prebuild command `npx expo prebuild --platform android --clean`
File `/android/app/build.gradle` will be updated with core library desugaring enabled
![image](https://github.com/user-attachments/assets/80c4edc0-5655-48fd-8ade-603e4c49bc35)

Hope this help!!

**Comment by long-vv** [comment]
@anhtuan219 It's working like a charm, thank you very much 💖
